include/crab/domains/mru_region_domain.hpp

```diff
@@ -0,0 +1,2270 @@
+    base_variable_vector_t old_base_vars, new_base_vars;
+    for (unsigned i = 0; i < from.size(); ++i) {
+      variable_t old_var = from[i];
+      auto search = m_var_map.find(old_var);
+      if (search == m_var_map.end()) { // NOT FOUND
+        continue;
+      }
+      ghost_variables_t old_ghost_vars = search->second;
+      // append base_var, [offset, size] into old_base_vars
+      old_ghost_vars.add(old_base_vars);
+      // remove from rev_var_map
+      old_ghost_vars.remove_rev_varmap(m_rev_var_map);
+      // remove from var_map
+      m_var_map.erase(old_var);
+
+      variable_t new_var = to[i];
+      // create/get new ghost variable for new_var
+      const ghost_variables_t &new_gvars = get_or_insert_gvars(new_var);
+      // append base_var, [offset, size] into new_base_vars
+      new_gvars.add(new_base_vars);
+    }
+
+    // rename mem domain
+    m_mem.rename(old_base_vars, new_base_vars);
+    // rename cache_lines domain
+    m_cache_lines.rename(old_base_vars, new_base_vars);
+    // rename regs domain
+    m_regs.rename(old_base_vars, new_base_vars);
+    // rename addrs domain
+    m_addrs.rename(old_base_vars, new_base_vars);
+
+    // rename the rest of environments
+    m_rgn_counting_dom.rename(from, to);
+    m_rgn_init_dom.rename(from, to);
+    if (!crab_domain_params_man::get().region_skip_unknown_regions()) {
+      m_rgn_type_dom.rename(from, to);
+    }
+    if (crab_domain_params_man::get().region_allocation_sites()) {
+      m_alloc_site_dom.rename(from, to);
+    }
+    if (crab_domain_params_man::get().region_deallocation()) {
+      m_rgn_dealloc_dom.rename(from, to);
+    }
+    if (crab_domain_params_man::get().region_tag_analysis()) {
+      m_tag_env.rename(from, to);
+    }
+  }
+
+  /* begin intrinsics operations */
+  void intrinsic(std::string name, const variable_or_constant_vector_t &inputs,
+                 const variable_vector_t &outputs) override {
+    //=================================================================//
+    //       Special intrinsics supported by the region domain
+    //=================================================================//
+    // ---DSA region analysis---
+    //      This analysis indicates which regions might belong to the
+    //      same memory object. The intrinstic is added only if object
+    //      has more than one field.
+    // TODO: need to support other analysis
+    auto error_if_not_rgn = [&name](const variable_t &var) {
+      if (!var.get_type().is_region()) {
+        CRAB_ERROR("Intrinsics ", name, " parameter ", var,
+                   " should be a region");
+      }
+    };
+
+    if (is_bottom()) {
+      return;
+    }
+
+    if (name == "regions_from_memory_object") {
+      // why pass to the outputs vector?
+      // https://github.com/seahorn/clam/blob/ee2f50723a40377864fee7618b90df92d1413a58/lib/Clam/CfgBuilder.cc#L4501
+      // pass region varaibles into cache_map
+      variable_t representative = outputs[0];
```

Comment:
so just use `inputs` to access to all region variables that are originated from the same dsa node.

Reply:
There is no representative. I just pass all the region variables. Not sure why you need a representative.

Reply:
You can use here the `union_find_domain`. This is like an union-find data structure but with lattice operations. Here, you take the first region variable and call `make` and then with the rest you call `join` (merge). This is how you would put all the region variables in the same equivalence class. 

Having said this, since we only have this kind of intrinsic at the beginning of each function joins shouldn't be a problem because both operands of the join should have the same equivalence classes since they are created at the beginning. So you can also have a vector of sets or something similar and in the binary lattice operations just report an error if both operands do not agree on the same equivalence classes.